test/mpi/threads/init/mult_session.c

```diff
@@ -0,0 +1,157 @@
+static void library_foo_finalize(int thread_idx, MPI_Session * p_session, MPI_Comm * p_comm);
+
+MTEST_THREAD_RETURN_TYPE library_foo_test(void *p)
+{
+    int rank, size;
+    int thread_idx = (int) (long) p;
+
+    MPI_Session lib_shandle = MPI_SESSION_NULL;
+    MPI_Comm lib_comm = MPI_COMM_NULL;
+    library_foo_init(thread_idx, &lib_shandle, &lib_comm);
+
+    MPI_Comm_size(lib_comm, &size);
+    MPI_Comm_rank(lib_comm, &rank);
+
+    thread_errs[thread_idx] += MTestTestIntracomm(lib_comm);
+
+    library_foo_finalize(thread_idx, &lib_shandle, &lib_comm);
+
+    return MTEST_THREAD_RETVAL_IGN;
+}
+
+static void library_foo_init(int thread_idx, MPI_Session * p_session, MPI_Comm * p_comm)
+{
+    int rc, flag;
+    int ret = 0;
+    const char *pset_name;
+    char out_value[100];        /* large enough */
+
+    /* Let's test both WORLD and SELF. e.g. with 4 threads, thread 2 will run on SELF */
+    if (thread_idx % 2) {
+        pset_name = "mpi://WORLD";
+    } else {
+        pset_name = "mpi://SELF";
+    }
+
+    MPI_Group wgroup = MPI_GROUP_NULL;
+    rc = MPI_Session_init(MPI_INFO_NULL, MPI_ERRORS_RETURN, p_session);
+    if (rc != MPI_SUCCESS) {
+        thread_errs[thread_idx]++;
+        printf("MPI_Session_init failed in thread %d\n", thread_idx);
+        goto fn_exit;
+    }
+
+    /* check we got thread support level foo library needs */
+    MPI_Info tinfo = MPI_INFO_NULL;
+    const char mt_key[] = "mpi_thread_support_level";
+    const char mt_value[] = "MPI_THREAD_MULTIPLE";
+    rc = MPI_Session_get_info(*p_session, &tinfo);
+    if (rc != MPI_SUCCESS) {
+        thread_errs[thread_idx]++;
+        goto fn_exit;
+    }
+
+    MPI_Info_get(tinfo, mt_key, sizeof(out_value), out_value, &flag);
+    if (flag != 1) {
+        thread_errs[thread_idx]++;
+        printf("Could not find key %s\n", mt_key);
+        goto fn_exit;
+    }
+    if (strcmp(out_value, mt_value)) {
+        thread_errs[thread_idx]++;
+        printf("Did not get thread multiple support, got %s\n", out_value);
+        goto fn_exit;
+    }
+
+    /* create a group from the WORLD process set */
+    rc = MPI_Group_from_session_pset(*p_session, pset_name, &wgroup);
+    if (rc != MPI_SUCCESS) {
+        thread_errs[thread_idx]++;
+        printf("MPI_Group_from_session_pset failed in thread %d\n", thread_idx);
+        goto fn_exit;
+    }
+
+    /* get a communicator */
+    char string_tag[20];
+    sprintf(string_tag, "thread %d", thread_idx);
+    rc = MPI_Comm_create_from_group(wgroup, string_tag, MPI_INFO_NULL, MPI_ERRORS_RETURN, p_comm);
+    if (rc != MPI_SUCCESS) {
+        thread_errs[thread_idx]++;
+        printf("MPI_Comm_create_from_group failed in thread %d\n", thread_idx);
+        goto fn_exit;
+    }
+
+    /* free group, library doesn’t need it. */
+  fn_exit:
+    MPI_Group_free(&wgroup);
+    if (tinfo != MPI_INFO_NULL) {
+        MPI_Info_free(&tinfo);
+    }
+    if (ret != 0) {
+        MPI_Session_finalize(p_session);
+    }
+}
```

Comment:
`ret != 0` might need to be replaced by `rc != MPI_SUCCESS `

Reply:
In case `library_foo_init` fails at some point, should we do not continue in `library_foo_test` as session has already been finalized?

Reply:
sure.

Reply:
Both fixed. `library_foo_init` now return `bool`.